### PR TITLE
Unsere loesung fix

### PR DESCRIPTION
**Pull request by lpdon**
Funktioniert noch nicht, aber zumindest läuft. Fehler war outputActivation immer = softmax

### CODE REVIEW

File: src/model/mlp.py [hunk 1]
```diff
@@ -60,9 +58,11 @@ def __init__(self, train, valid, test, layers=None, inputWeights=None,
             self.loss = DifferentError()
         elif loss == 'absolute':
             self.loss = AbsoluteError()
+        elif loss == 'crossentropy':
+            self.loss = CrossEntropyError()
         else:
             raise ValueError('There is no predefined loss function ' +
-                             'named ' + str)
```
Comment (m1cm1c) [inline src/model/mlp.py:65]: Wtf? Haben wir das vorher verkackt, oder war das schon so?

File: src/model/mlp.py [hunk 2]
```diff
@@ -136,26 +133,23 @@ def _compute_error(self, target):
...
         tempDerivatives = None
         for layer in reversed(self.layers):
             if layer == self._get_output_layer():
-
                 layer.computeDerivative(self.loss.calculateDerivative(label, layer.outp), 1.0)
                 tempWeights = layer.weights
                 tempDerivatives = layer.deltas
             else:
-                layer.computeDerivative(tempDerivatives, tempWeights)
+                layer.computeDerivative(tempDerivatives, tempWeights[1:])
```
Comment (m1cm1c) [inline src/model/mlp.py:150]: Ist das, damit der Bias bei der Ableitung nicht berücksichtigt wird? Falls ja: Wieso sollte das so sein? Dazu finde ich nichts in den Folien.

File: src/model/mlp.py (hunk 1 above)
Comment (lpdon) [reply src/model/mlp.py:65]: War schon falsch auf Ex3

Comment (lpdon) [reply src/model/mlp.py:65]: War schon falsch auf Branch Ex3.

File: src/model/mlp.py (hunk 2 above)
Comment (lpdon) [reply src/model/mlp.py:150]: Ja. Wir haben eine Ableitung pro Neuron in der nächsten Schichte. Für die aktuelle Schicht, für jedes Neuron, haben wir den Fehler aller Neuronen der nächsten Schicht, die sich mit diesem Neuron der aktuellen Schicht verbinden. Das ist das "Downstream" von der Folien. 

Mit Zahlen: 

Hidden Layer: 128 Neuronen
Output Layer: 10 Neuronen

das entspricht: 
- 129x10 Gewichte in Output Layer (jedes Neuron, 128 Verbindungen mit der Hidden Layer + Bias)
- 10x1 Deltas in Output Layer(1 Delta pro Neuron) 
- Deltas der Hidden Layer müssen 128 Elemente haben (1 Delta pro Neuron). D.h: deltas_hidden = weights_output (ohne Bias, da der Bias keine Verbindung zwischen Schichten ist) * deltas_output

Comment (m1cm1c) [reply src/model/mlp.py:150]: Klar kann man beim Bias nicht weiter propagieren, aber er muss doch auch aktualisiert werden.

Comment (lpdon) [reply src/model/mlp.py:150]: Er wird auch aktualisiert. Bei dem Update wird er aktualisiert da er auch ein Input des Neurons ist. Der Bias hat nur keinen Einfluss auf die Deltas.

image

Comment (lpdon) [reply src/model/mlp.py:150]: Weitere Anmerkungen? Sonst würde ich sagen dass wir die Änderungen mergen und mit dem Bericht anfangen.
